Add full support for Tuya BAC-002-ALZB (schedule_text, improved off behavior, calibration -9/+9)

kuposcar opened the pull request:
What does this PR do?

This PR adds full native support for the Tuya BAC-002-ALZB FCU thermostat.

Improvements included

Adds single-line unified weekly schedule_text (12 segments), readable and writable from Home Assistant or MQTT.

Fixes the long-standing issue where the thermostat would not turn off on first attempt.

Adds correct temperature calibration range −9 to +9°C, useful when using an external sensor.

Uses the modernExtend tuyaBase with forceTimeUpdates to ensure proper time sync.

Provides stable system_mode → Tuya DP mapping and state consolidation.

Adds whiteLabel and full fingerprint support for _TZE200_dzuqwsyg and _TZE204_dzuqwsyg.

Compatibility requirements

Works on Zigbee2MQTT 2.7.1 or newer (uses modernExtend and unified Tuya datapoints).

Notes

Tested extensively on the actual device.
Provides schedule parsing identical to older versions but in a cleaner, safer, more automation-friendly format.

kuposcar commented:
I’ve already identified and fixed the issue.
It seems to originate from Home Assistant.

I’ll open a PR as soon as I can. In the meantime, I’m sharing the external converter in case anyone wants to use it until the next release.

I’ve tested the external converter myself and also shared it with a few other users, and so far it appears to be working well.

I’ve also updated the deadzone range to go from 0 to 5; previously it started at 1.

The file is provided in .mjs format so it can be merged directly later on.

[custom_bac002_2.7.1.zip](https://github.com/user-attachments/files/24441992/custom_bac002_2.7.1.zip)

robvanoostenrijk commented:
@kuposcar,

Thanks for improving the BAC-002-AZLB / BAC-003 integration and providing the external converter for validation. 

I've reviewed your changes:

- Off Modifications
   
   I can see you are now turning the device on, in order to turn it off 120ms later.
   On my devices this leads to an unnecessary backlight flash (screen lighting up) when turning off.
   Direct off works fine on both BAC-003 and BAC-002-ALZB. 
   
   Are you sure the change you made is not working around another automation in a downstream system conflicting with this mode change? The best way to test is to disconnect Zigbee2MQTT from downstream and ensure you don't have something like Zigbee2MQTT automations add-on loaded.

- Device Names

   From the information I can see on the internet:
   `_TZE204_dzuqwsyg`: `BAC-003` 
   `_TZE200_dzuqwsyg`: `BAC-002-ALZB`

   **Action**: Updated the whitelabel for BAC-003.
   
- Schedule Text

  Good change, it will indeed made automation values easier to set, which is the most likely interaction method.

- Deadzone Temperature

  It was previously limited to -3..3 because because these are the maximum values selectable in the on device maintenance menu (tested on `_TZE204_dzuqwsyg` and `_TZE200_dzuqwsyg`). The manual for the WiFi one I found online does indeed say -9..+9.
I've confirmed however that -9..+9 works, however if you set -9 from Zigbee2MQTT, in the on-device menu you can then continue past the built-in block at -3 to -A, -B, -C (hex values). I'm not sure if thats healthy and could corrupt the device memory, particularly if continuing into double byte values.

  Hidden menu settings for reference:
  <img width="332" height="342" alt="image" src="https://github.com/user-attachments/assets/4fd35cd3-db09-4410-9207-d54458d207ed" />

- System Mode (Cooling Only)

  I run these devices in a hot climate. As such they are 2 pipe models (in cold, out cold), compared to 4 pipe models which have (in hot, in cold, out cold, out hot).  When using your converter and integrating it with Google Home, it gets confused and shows only heat mode.
 
  <img width="540" height="391" alt="image" src="https://github.com/user-attachments/assets/bee2df2c-916e-45fd-9724-841f0840ce7a" />
 
   I can see that the device specific configuration for it is still in place, but the code handling the setting has maybe been accidentally removed.
 
   **Action**:  Added back in the system mode modification in expose for 2-pipe mode.

- System Mode (Add Off)
   On these devices, the system mode datapoint does not include off. Thats why previously I created a `virtual` off by adding it in the system mode and sending the right data point commands.
   However, some using downstream integrations wanted the off in system mode to be optional, thats why there was a configuration for it. I can see you left the code for it in datapoint 1 handling, so maybe this was an oversight.
  
   The option splits the off function back into the original state property. 

   **Action**: I've added it back in expose for now. However if the community does not require this anymore it could be removed.

[custom_bac002_2.7.1_mod.zip](https://github.com/user-attachments/files/24444859/custom_bac002_2.7.1_mod.zip)

kuposcar commented:
Thanks for the detailed review, I’ll take a look at it.

The reason for the power-on before power-off is that on the BAC-002 units I tested, the device would not turn off when sending an OFF command directly, as it did not “wake up” first.

Regarding OFF handling, it’s a bit tricky because by default ON/OFF is independent from the system mode, and including it inside the mode causes issues. In the previous external converter I made, the main complaints were precisely that ON/OFF was independent from the mode, matching the original Tuya behavior.

Just to confirm: when you mention deadzone, you are referring to the temperature calibration/compensation, correct?

For reference, I’ve been using calibration values in the -9 to +9 range on my devices for about a year without observing any issues or unexpected behavior.

If your converter works well for everyone, we can base the merge on your implementation.

Thanks a lot.

robvanoostenrijk commented:
> Thanks for the detailed review, I’ll take a look at it.
> 
> The reason for the power-on before power-off is that on the BAC-002 units I tested, the device would not turn off when sending an OFF command directly, as it did not “wake up” first.

Based on your feedback, I've added a device specific setting to do this: "Wake before power off".
When this is enabled it will send the power-on command and wait 120ms to send the power-off command as per your logic.
If disabled, it just turns the device off as before.

> Regarding OFF handling, it’s a bit tricky because by default ON/OFF is independent from the system mode, and including it inside the mode causes issues.

That's why the device-specific setting is there. If you want a separate on/off switch entity, you flip that setting in Z2M.
The combined system mode is required for downstream systems such as Alexa and Google Home, these do not understand a separate on/off switch entity for an Airconditioning control unit.

> Just to confirm: when you mention deadzone, you are referring to the temperature calibration/compensation, correct?
Sorry, temperate calibration indeed. Just highlighting that while -/+9 works it goes beyond what the on-device menu allows.

> For reference, I’ve been using calibration values in the -9 to +9 range on my devices for about a year without observing any issues or unexpected behavior.
Anything over -/+3 seems quite a large measurement deviation. But lets keep this in then, its bound-limited by Z2M regardless.

> If your converter works well for everyone, we can base the merge on your implementation.

Let's merge this all together and re-submit the pull request. All pre-existing use-cases are now covered.

Attaching the updated version with a device-setting for "Wake before poweroff":
[custom_bac002_2.7.1_mod_2.zip](https://github.com/user-attachments/files/24448819/custom_bac002_2.7.1_mod_2.zip)

kuposcar commented:
> > Thanks for the detailed review, I’ll take a look at it.
> > The reason for the power-on before power-off is that on the BAC-002 units I tested, the device would not turn off when sending an OFF command directly, as it did not “wake up” first.
> 
> Based on your feedback, I've added a device specific setting to do this: "Wake before power off". When this is enabled it will send the power-on command and wait 120ms to send the power-off command as per your logic. If disabled, it just turns the device off as before.
> 
> > Regarding OFF handling, it’s a bit tricky because by default ON/OFF is independent from the system mode, and including it inside the mode causes issues.
> 
> That's why the device-specific setting is there. If you want a separate on/off switch entity, you flip that setting in Z2M. The combined system mode is required for downstream systems such as Alexa and Google Home, these do not understand a separate on/off switch entity for an Airconditioning control unit.
> 
> > Just to confirm: when you mention deadzone, you are referring to the temperature calibration/compensation, correct?
> > Sorry, temperate calibration indeed. Just highlighting that while -/+9 works it goes beyond what the on-device menu allows.
> 
> > For reference, I’ve been using calibration values in the -9 to +9 range on my devices for about a year without observing any issues or unexpected behavior.
> > Anything over -/+3 seems quite a large measurement deviation. But lets keep this in then, its bound-limited by Z2M regardless.
> 
> > If your converter works well for everyone, we can base the merge on your implementation.
> 
> Let's merge this all together and re-submit the pull request. All pre-existing use-cases are now covered.
> 
> Attaching the updated version with a device-setting for "Wake before poweroff": [custom_bac002_2.7.1_mod_2.zip](https://github.com/user-attachments/files/24448819/custom_bac002_2.7.1_mod_2.zip)


Thanks a lot for the work, I really appreciate the effort you’ve put into this 👍

I’ve been testing this myself and together with a few other users. Overall, it works well, but we’ve noticed a couple of practical issues.

The main one is that the wake_before_power_off on/off switch does not appear in the UI. This means users would need to touch YAML or rely on a modified external converter to control it, which could be a problem for many users.

Based on your feedback and your external implementation, I tried an alternative approach: instead of powering on before powering off, I’m now sending two consecutive power-off commands. With this approach, the thermostat consistently turns off, and it does so without the LED flashing that was mentioned in earlier comments.

Another issue I personally ran into with this converter was that, once the device was turned off, it would not always turn back on. To address this, I added a small piece of logic where, if the device is off, it briefly powers on first and then switches to the requested mode. With this change in place, the behavior seems stable.

In any case, if the device is off and you set any system mode, the LED will turn on anyway, so I don’t think this causes any real negative side effects.

If you can confirm that this external behaves correctly for you as well, I think we’re in a good position to move forward and use this as the final solution.
[custom_bac002_2.7.1_mod_3.zip](https://github.com/user-attachments/files/24472154/custom_bac002_2.7.1_mod_3.zip)
